File: Leetcode657. Robot return to origin/Solution.java
class Solution {
    public boolean judgeCircle(String moves) {
        int x=0, y=0;
        for(char move : moves.toCharArray()) {
            if(move == 'U') y--;
            else if (move == 'D') y++;
            else if (move == 'L') x--;
            else if(move == 'R') x++;
        }
        return x==0 && y==0;
    }
}


// runtime - 2ms
class Solution {
    public boolean judgeCircle(String moves) {
        int[] ch = new int[26];
        for (char move : moves.toCharArray()){
            ch[move - 'A']++;
        }
        return ch['U' - 'A'] == ch['D' - 'A'] && 
                ch['L' - 'A'] == ch['R' - 'A'];
    }
}


// runtime - 3ms
class Solution {
    public boolean judgeCircle(String moves) {
        int[] ch = new int[26];
        for (char move : moves.toCharArray()){
            ch[move - 'A']++;
        }
        return ch['U' - 'A'] == ch['D' - 'A'] && 
                ch['L' - 'A'] == ch['R' - 'A'];
    }
}


// runtime - 4ms
class Solution {
    public boolean judgeCircle(String moves) {
        int right=0;
        int up=0;
        for(final char c:moves.toCharArray()) {
            switch(c) {
                case 'R':
                    right++;
                    break;
                case 'L':
                    right--;
                    break;
                case 'U':
                    up++;
                    break;
                case 'D':
                    up--;
                    break;
            }
        }
        return right==0 && up==0;
    }
}

File: Leetcode657. Robot return to origin/Solution.py
class Solution(object):
    def judgeCircle(self, moves):
        x = y = 0
        for move in moves:
            if move == 'U': y -= 1
            elif move == 'D': y += 1
            elif move == 'L': x -= 1
            elif move == 'R': x += 1

        return x == y == 0


# runtime - 0ms
class Solution:
    def judgeCircle(self, moves: str) -> bool:
        return moves.count('U')==moves.count('D') and moves.count('L')==moves.count('R')


# runtime - 7ms
class Solution:
    def judgeCircle(self, moves: str) -> bool:
        cnt = Counter(moves)
        return cnt['U'] == cnt['D'] and cnt['L'] == cnt['R']


# runtime - 10ms
class Solution:
    def judgeCircle(self, moves: str) -> bool:
        if len(moves) % 2:
            return False
        c = Counter(moves)
        l = c.get('L', 0)
        r = c.get('R', 0)
        u = c.get('U', 0)
        d = c.get('D', 0)
        return l == r and u == d
